### PR TITLE
feat: Add support for push notifications via `ParsePush`, `ParseNotification` 

### PR DESCRIPTION
**Pull request by mbfakourii**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

## Issue
Closes: #913 and #742

## Approach
This feature is implemented based on the logic of Parse Android

## Tasks

- [ ] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)

### CODE REVIEW

**Comment by mtrezza** [comment]
Is this actually a new feature or a bug fix for an existing feature?

**Comment by mbfakourii** [comment]
> Is this actually a new feature or a bug fix for an existing feature? Was there no push notification support at all so far?

Such a feature did not exist before and is a new feature

File: packages/flutter/PUSH.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Push Notifications
+
+Push notifications are a great way to keep your users engaged and informed about your app. You can reach your entire user base quickly and effectively. This guide will help you through the setup process and the general usage of Parse to send push notifications.
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:3]:
```suggestion
Push notifications are a great way to keep your users engaged and informed about your app. You can reach your user base quickly and effectively. This guide will help you through the setup process and the general usage of Parse Platform to send push notifications.
```

File: packages/flutter/PUSH.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+To activate and implement Push Notifications in Parse Server, check [this page](https://docs.parseplatform.org/parse-server/guide/#push-notifications)
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:5]:
```suggestion
To configure push notifications in Parse Server, check out the [push notification guide](https://docs.parseplatform.org/parse-server/guide/#push-notifications).
```

File: packages/flutter/PUSH.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Put the following code after the above codes
+```dart
+FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
+```
+
+Put the following function in your codes
+```dart
+Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);
+```
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:30]:
```suggestion
## Installation

1. Install [Firebase Core](https://firebase.flutter.dev/docs/overview) and [Cloud Messaging](https://firebase.flutter.dev/docs/messaging/overview). For more details review the [Firebase Core Manual](https://firebase.flutter.dev/docs/manual-installation/).

2. Add the following code after `Parse().initialize(...);`:

  ```dart
  ParsePush.instance.initialize(FirebaseMessaging.instance);
  FirebaseMessaging.onMessage.listen((message) => ParsePush.instance.onMessage(message));
  ```

3. For you app to process push notification while in the background, add the following code:

  ```dart
  FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
  ```

  ```dart
  Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);
  ```

File: packages/flutter/PUSH.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+## Implemented example
+Your code should look like the following code
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:33]:
```suggestion
## Implementation Example

The following is a code example for a simple implementation of push notifications:
```

File: packages/flutter/PUSH.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  // initialize Parse Push
+  ParsePush.instance.initialize(FirebaseMessaging.instance);
+  FirebaseMessaging.onMessage
+      .listen((message) => ParsePush.instance.onMessage(message));
+  
+  // for run ParsePush in the background
+  FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
+
+  runApp(const MyApp());
+}
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:57]:
```suggestion
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  
  // Initialize Firebase Core
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  // Initialize Parse
  await Parse().initialize("applicationId", "serverUrl",
      clientKey: "clientKey", debug: true);

  // Initialize Parse push notifications
  ParsePush.instance.initialize(FirebaseMessaging.instance);
  FirebaseMessaging.onMessage
      .listen((message) => ParsePush.instance.onMessage(message));
  
  // Process push notifications while app is in the background
  FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);

  runApp(const MyApp());
}
```

File: packages/flutter/lib/src/notification/parse_notification.dart [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+class ParseNotification {
+  static final ParseNotification instance = ParseNotification._internal();
+  static String keyNotificationChannelName = "parse";
+
+  factory ParseNotification() {
+    return instance;
+  }
+
+  ParseNotification._internal() {
+    // initialize Awesome Notifications
```
Comment (mtrezza) [inline packages/flutter/lib/src/notification/parse_notification.dart:12]:
```suggestion
    // Initialize notifications helper package
```

File: packages/flutter/lib/src/push/parse_push.dart [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  static String keyPushType = 'pushType';
+
+  factory ParsePush() {
+    return instance;
+  }
+
+  ParsePush._internal();
+
+  Future<void> initialize(firebaseMessaging) async {
+    // get GCM token
```
Comment (mtrezza) [inline packages/flutter/lib/src/push/parse_push.dart:15]:
```suggestion
    // Get Google Cloud Messaging (GCM) token
```

File: packages/flutter/lib/src/push/parse_push.dart [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    firebaseMessaging.getToken().asStream().listen((event) async {
+      // set token in ParseInstallation
```
Comment (mtrezza) [inline packages/flutter/lib/src/push/parse_push.dart:17]:
```suggestion
      // Set token in installation
```

File: packages/flutter/lib/src/push/parse_push.dart [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  void _handlePush(String pushId, String timestamp, String channel,
+      Map<String, dynamic>? data) {
+    if (pushId.isEmpty || timestamp.isEmpty) {
+      return;
+    }
+
+    if (data != null) {
+      // show notification
```
Comment (mtrezza) [inline packages/flutter/lib/src/push/parse_push.dart:49]:
```suggestion
      // Show push notification
```

File: packages/flutter/lib/src/push/parse_push.dart [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  Future<void> unsubscribeFromChannel(String value) async {
+    sdk.ParseInstallation parseInstallation =
+        await sdk.ParseInstallation.currentInstallation();
+
+    await parseInstallation.unsubscribeFromChannel(value);
+  }
+
+  Future<List<dynamic>> getSubscribedChannels() async {
+    sdk.ParseInstallation parseInstallation =
+        await sdk.ParseInstallation.currentInstallation();
```
Comment (mtrezza) [inline packages/flutter/lib/src/push/parse_push.dart:70]: These seem to be really short line lengths; which makes this difficult to read. What line length have we set in the linter?

Comment (Nidal-Bakir) [reply packages/flutter/lib/src/push/parse_push.dart:70]: 80 char per line is the defalut. In the Flutter ecosystem, this is the standard line length.

Comment (mtrezza) [reply packages/flutter/lib/src/push/parse_push.dart:70]: Seems outdated in the age of widescreen monitors, but fine.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Please add changelog entry and version bump; is this then ready for merge?

File: packages/flutter/PUSH.md (hunk 3 above)
Comment (mbfakourii) [reply packages/flutter/PUSH.md]: Do you think the Installation section was too short? Of course it is understandable!

File: packages/flutter/lib/src/push/parse_push.dart (hunk 8 above)
Comment (mbfakourii) [reply packages/flutter/lib/src/push/parse_push.dart]: The concept of installation cannot be understood where it is being set.
Are you sure about this?

**Comment by mbfakourii** [comment]
> Please add changelog entry and version bump; is this then ready for merge?

I tested it several times and it worked fine.
@Nidal-Bakir 
What is your opinion? Is it ready for merge?

File: packages/flutter/PUSH.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+3. For you app to process push notification while in the background, add the following code:
+```dart
+FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
+```
+
+```dart
+Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);
+```
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:26]: Why has the indentation been removed?

File: packages/flutter/PUSH.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+1. Install [Firebase Core](https://firebase.flutter.dev/docs/overview) and [Cloud Messaging](https://firebase.flutter.dev/docs/messaging/overview). For more details review the [Firebase Core Manual](https://firebase.flutter.dev/docs/manual-installation/).
+
+2. Add the following code after `Parse().initialize(...);`:
+
+```dart
+ParsePush.instance.initialize(FirebaseMessaging.instance);
+
+FirebaseMessaging.onMessage.listen((message) => ParsePush.instance.onMessage(message));
+```
```
Comment (mtrezza) [inline packages/flutter/PUSH.md:17]: Why has the indentation been removed?

File: packages/flutter/lib/src/push/parse_push.dart [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+  factory ParsePush() {
+    return instance;
+  }
+
+  ParsePush._internal();
+
+  /// initialize ParsePush
+  ///
+  /// On web, a [vapidKey] is required.
```
Comment (mtrezza) [inline packages/flutter/lib/src/push/parse_push.dart:17]:
```suggestion
  /// Initialize ParsePush; for web a [vapidKey] is required.
```

**Comment by mbfakourii** [comment]
> Please add changelog entry and version bump; is this then ready for merge?

added

File: packages/flutter/CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,9 @@
+## [5.1.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.0.1...flutter-5.1.0) (2023-05-22)
+
+### Features
+
+* Add `ParsePush` and `ParseNotification` ([#914](https://github.com/parse-community/Parse-SDK-Flutter/pull/914))
```
Comment (mtrezza) [inline packages/flutter/CHANGELOG.md:5]:
```suggestion
* Add support for push notifications via `ParsePush`, `ParseNotification` ([#914](https://github.com/parse-community/Parse-SDK-Flutter/pull/914))
```

**Review by mtrezza — APPROVED** [review]
Looks good!

**Comment by mbfakourii** [comment]
It seems that we have an error in CI and the [codecov section](https://github.com/parse-community/Parse-SDK-Flutter/actions/runs/5047305654/jobs/9053992456?pr=914)

```
Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.
```

**Comment by mbfakourii** [comment]
I checked CI, it is using a file in this [line](https://github.com/parse-community/Parse-SDK-Flutter/blob/f8942ac0a7e5139930c68c870aeb5547e2330100/.github/workflows/ci.yml#LL148C18-L148C58) that we don't have in this repository

**Comment by mtrezza** [comment]
Yes, these codedov timeouts occur sometimes, it's an API limit, no issue on our side.

**Comment by cool2apps** [comment]
Is there any documentation how to use ParsePush and ParseNotification?

**Pull request merged**